gaps/gap-24_agreement_permissions_management/gap-24_agreement_permissions_management.md

```diff
@@ -0,0 +1,198 @@
+
+A coherent permissions management model may also enable previously unfeasible scenarios where eg. a "broker" Requestor node negotiates Agreements with Providers and then sells them to intrested customers, who then execute the Agreement by launching Activities and accepting Invoices/DebitNotes.
+
+## Specification
+The technical specification should describe the syntax and semantics of any new feature. 
+
+### Scope
+
+Following entities are in scope of Agreement's permissions management: 
+- Agreement
+- Activity
+- Invoice
+- DebitNote
+- Payment
+- ExecBatch
+
+All other Golem entities (as exposed by Golem low-level APIs) are not covered by Agreement Permissions mechanism.
+
+### Roles
+From the point of view of Agreement permissions, following actor roles are relevant:
+
+- *Owner* - Requestor and Provider which signed the Agreement as a result of negotiations are always the Owners of the Agreement (and all related entities). 
+- *Grantee* - any other (non-Owner) node/ID, which has been granted permissions and is included in Agreement's Access Control List.
+
+### Permissions catalogue
+
+A **permission** indicates an ability to perform a specific action which refers to an Agreement or any related entity. For the purposes of rights management, the permissions are indicated via literals, which include permission's category and specific action - where actions map to respective low-level Golem REST API operations, as implemented. Following table summarizes permissions which are covered by this GAP:
+
+|Permission|Comments|
+|-|-|
+| req:getAgreement | |
+| req:getAgreementPermissions | |
+| req:collectAgreementEvents | |
+| req:getActivityState | |
+| req:getActivityUsage | |
+| req:getRunningCommand | |
+| req:getExecBatchResults | |
+| req:getDebitNotes | |
+| req:getDebitNote | |
+| req:getPaymentsForDebitNote | |
+| req:getDebitNoteEvents | |
+| req:getInvoices | |
+| req:getInvoice | |
+| req:getPaymentsForInvoice | |
+| req:getInvoiceEvents | |
+| req:getPayments | |
+| req:terminateAgreement | |
+| req:setAgreementPermissions | See: [Transitive permissions](#transitive-permissions) below |
+| req:createActivity | |
+| req:exec | |
+| req:callEncrypted | |
+| req:destroyActivity | |
+| req:acceptDebitNote | See: [Accept Invoice and Debit Note logic](#accept-invoice-and-debit-note-logic) below |
+| req:rejectDebitNote | |
+| req:acceptInvoice | See: [Accept Invoice and Debit Note logic](#accept-invoice-and-debit-note-logic) below |
+| req:rejectInvoice | |
```

Comment:
Permissions could be hierarchical. For example it doesn't makes not so much sense to have rights to accept DebitNotes but not accept Invoices. It is strange not to have rights to reject Invoice when you can accept.

Of course I can imagine scenarios, when it is needed, but in general I would have hierarchy like this:
| req:payments:DebitNote:accept | |
| req:payments:DebitNote:reject | |
| req:payments:Invoice:accept | |
| req:payments:Invoice:reject | |
So you can for example give permissions to `req:payments`, which means you have all sub-permissions in hierarchy.

In other areas this hierarchy could be different. For example probably in general in is not harmful to have read permissions, so for `get..`
| req:get:payments:Invoices | |
| req:get:payments:DebitNoteEvents | |
| req:get:activity:state | |
| req:get:activity:usage | |
This way you can give all `read-only` permissions at once.

Reply:
I would prefer to stick with a flat permissions list, directly mapped to API calls - unless there is a very strong reason to make this more complex.